# Exploring eBay Car Sales Data with NumPy & Pandas
This project will be a good exercise in utilizing everything we have learned thus far with NumPy and pandas. We will be cleaning up and aggregating a lot of data in this project, and we'll be doing it by using different indexing methods, boolean masks, and other data cleaning techniques. We have a 

The data dictionary provided with data is as follows:

* dateCrawled - When this ad was first crawled. All field-values are taken from this date.
* name - Name of the car.
* seller - Whether the seller is private or a dealer.
* offerType - The type of listing
* price - The price on the ad to sell the car.
* abtest - Whether the listing is included in an A/B test.
* vehicleType - The vehicle Type.
* yearOfRegistration - The year in which the car was first registered.
* gearbox - The transmission type.
* powerPS - The power of the car in PS.
* model - The car model name.
* kilometer - How many kilometers the car has driven.
* monthOfRegistration - The month in which the car was first registered.
* fuelType - What type of fuel the car uses.
* brand - The brand of the car.
* notRepairedDamage - If the car has a damage which is not yet repaired.
* dateCreated - The date on which the eBay listing was created.
* nrOfPictures - The number of pictures in the ad.
* postalCode - The postal code for the location of the vehicle.
* lastSeenOnline - When the crawler saw this ad last online.
* The aim of this project is to clean the data and analyze the included used car listings. You'll also become familiar with some of the unique benefits jupyter notebook provides for pandas.

Let's start by importing the libraries we need and reading the dataset into pandas.

In [1]:
import numpy as np
import pandas as pd
import re

#read in our data to a pandas dataframe
#NOTE: pd.read_csv method needs the r for it to work properly due to file location
autos = pd.read_csv(r'my_datasets\auto.csv', encoding='Latin-1')

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [3]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

As we can see from the above, we have a ton of data, and some of it is in German. We have many columns containing strings and int datatypes. There are lots of inconsistencies and abnormalities within this dataset. There are a lot of null-values in our data as well. Before we do anything serious, let's organize our column names a bit first. We'll turn all of the camelCase column names into snake_case, as well as rename some of the columns to make them a bit more descriptive and logical.

In [4]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [5]:
def camel_to_snake(text): #i found this online, and it works
    import re
    str1 = re.sub('(.)([A-Z][a-z]+)', r'\1_\2', text)
    return re.sub('([a-z0-9])([A-Z])', r'\1_\2', str1).lower()
    
def replace_col(col):
    col = col.replace("year_of_registration","registration_year")
    col = col.replace("month_of_registration","registration_month")
    col = col.replace("not_repaired_damage","unrepaired_damage")
    col = col.replace("nr_of_pictures","num_photos")
    col = col.replace("date_created","ad_created")
    col = col.replace("kilometer","odometer_km")
    return col

temp_col = autos.columns
new_columns = []
for c in temp_col:
    c = camel_to_snake(c)
    c = replace_col(c)
    new_columns.append(c)
    
autos.columns = new_columns
autos.head() #test

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

So now we have our columns ready to go. We changed a couple of column names completely, purely to make them shorter and more consistent with the others. Since this came from the German eBay, perhaps some things have gotten a bit lost in translation as well. For example, "notRepairedDamage" can be a bit confusing, so we've changed it to "unrepaired_damage".

Now we're going to explore the data a bit more to see if there's anything we can simplify. For example, text columns where all or almost all of the values are the same. We can simply remove these as they don't add much for analysis purposes. Also any numeric data that is stored as text can be edited, cleaned, and converted.

In [6]:
col_to_explore = ['date_crawled',
                  'name',
                  'seller',
                  'offer_type',
                  'price',
                  'abtest',                  
                  'vehicle_type',
                  'registration_year',
                  'gearbox',
                  'power_ps',
                  'model',
                  'odometer',
                  'registration_month',
                  'fuel_type',
                  'brand',
                  'unrepaired_damage',
                  'ad_created',
                  'num_photos',
                  'postal_code',
                  'last_seen']

#loop through the above columns and look at the value counts of their unique contents
#we're looking for the columns that contain mostly the same values
for col in col_to_explore:
    print(autos[col].value_counts(),'\n')

2016-03-21 20:37:19    3
2016-03-30 17:37:35    3
2016-03-09 11:54:38    3
2016-03-23 18:39:34    3
2016-03-29 23:42:13    3
                      ..
2016-03-31 11:57:12    1
2016-03-05 22:39:50    1
2016-03-07 11:52:22    1
2016-03-09 21:45:42    1
2016-03-07 10:40:55    1
Name: date_crawled, Length: 48213, dtype: int64 

Ford_Fiesta                                                       78
Volkswagen_Golf_1.4                                               75
BMW_316i                                                          75
BMW_318i                                                          72
Volkswagen_Polo                                                   72
                                                                  ..
VW_Caddy_5_Sitzer__wenig_Kilometer_                                1
Volkswagen_POLO_6n_1.4                                             1
Mercedes_Benz_SL_55_AMG_Automatik_Vmax_300_Garantie_bis_6.5.16     1
Mercedes_Benz_CLK_Coupe_230_Kompressor_Master_Edition 

The main ones to look at are 'seller', 'offer_type', and 'num_photos'. The first two contain almost entirely the same value, while the third contains nothing but zeroes. So we can safely drop these columns from our data.

In [7]:
autos = autos.drop(['seller','offer_type','num_photos'], axis=1)

Next we are going to look at columns that contain numeric data stored as text. We will clean them up and convert them to the appropriate data type. By looking at the above again, we can see that the 'price' and 'odometer' columns can be simplified by removing all the non-digit characters.

In [8]:
autos['price'] = (autos['price']
                  .str.replace('$','')
                  .str.replace(',','')
                  .astype(int))

autos['price'].head()

<ipython-input-8-194571f1379e>:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will*not* be treated as literal strings when regex=True.
  autos['price'] = (autos['price']


0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int32

Now we will clean and convert the odometer column.

In [9]:
autos['odometer'] = (autos['odometer']
                  .str.replace('km','')
                  .str.replace(',','')
                  .astype(int))

autos.rename({'odometer':'odometer_km'}, axis=1, inplace=True)
autos['odometer_km'].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int32

## Exploring Mileage and Price

In [10]:
print(autos['odometer_km'].value_counts())

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64


This looks pretty reasonable, as all of the values are nice, rounded numbers (probably due to a pre-set on the  website). It does look like this particular set of car listings has more cars with higher mileages though.

In [11]:
print(autos['price'].unique().shape)
print(autos['price'].describe())
autos['price'].value_counts().head(20)

(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
999      434
750      433
900      420
650      419
850      410
700      395
4500     394
300      384
2200     382
950      379
Name: price, dtype: int64

This is a little odd, as the maximum value in the price column is over 1,000,000,000. Looking at the value_counts(), we can see that there are 1,400+ cars listed with a price of 0. Perhaps we can remove these from our data set. Also, we can see that most of the prices are also rounded to an even number. Perhaps this is humans' tendency to round up/down.

In [12]:
autos['price'].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

We can see from the data above that we have a couple anomalies. First, the most expensive car is almost one hundred million dollars. The following several listings are also astronomically high numbers for cars listed on an auction website. There's also a considerable jump in price after 350k. We can probably remove these outliers, and only keep cars that are less than 350k. Because it is an auction website, it is plausible that starting bids can be as low as one dollar, so we will keep everything between 1-350k.

In [13]:
#This could vary, depending on what you deem is a realistic price!
autos = autos[autos['price'].between(1,350000)]
autos["price"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

## Exploring the date columns¶
There are a number of columns with date information:

* date_crawled
* registration_month
* registration_year
* ad_created
* last_seen

These are a combination of dates that were crawled, and dates with meta-information from the crawler. The non-registration dates are stored as strings.

We'll explore each of these columns to learn more about the listings.

In [14]:
autos[['date_crawled', 'ad_created', 'last_seen']].head()

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [15]:
print(autos['date_crawled'].describe(),'\n')

#sorts them by date (oldest to newest)
print('Percentages by date - oldest to newest:')
print((autos['date_crawled']
       .str[:10]
       .value_counts(normalize=True, dropna=False)
       .sort_index(),'\n'))

#sorts them based on distribution (largest to smallest)
print('Percentages by size - largest to smallest:')
print((autos['date_crawled']
       .str[:10]
       .value_counts(normalize=True, dropna=False)
       .sort_values(ascending=False)))

count                   48565
unique                  46882
top       2016-04-04 16:40:33
freq                        3
Name: date_crawled, dtype: object 

Percentages by date - oldest to newest:
(2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, d

It seems like the site was crawled daily, as the distribution of listings is fairly uniform. 

In [16]:
print((autos['last_seen']
      .str[:10]
      .value_counts(normalize=True, dropna=False)
      .sort_index()))

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64


The crawler recorded the date it last saw any listing, which allows us to determine on what day a listing was removed, presumably because the car was sold.

The last three days contain a disproportionate amount of 'last seen' values. Given that these are 6-10x the values from the previous days, it's unlikely that there was a massive spike in sales, and more likely that these values are to do with the crawling period ending and don't indicate car sales.

In [17]:
print((autos['ad_created']
      .str[:10]
      .value_counts(normalize=True, dropna=False)
      .sort_index()))
print(autos['ad_created'].unique().shape)

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64
(76,)


Most of the ads were created within 1-2 months of the listing, but a few outliers were created as old as 9 months before being listed.

In [18]:
#shows us what the registration year data looks like
print(autos['registration_year'].describe())

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64


This is a strange one because the minimum and maximum years show impossible dates for the cars to be registered. The minimum is 1000, which is clearly before cars were even invented. The max shows 9999, which is 7000+ years away. We can examine this a bit further, and try to only keep the cars that were registered within a more normal timeframe.

In [19]:
print(autos['registration_year']
     .value_counts(normalize=True, dropna=False)
     .sort_index())

1000    0.000021
1001    0.000021
1111    0.000021
1800    0.000041
1910    0.000103
          ...   
5911    0.000021
6200    0.000021
8888    0.000021
9000    0.000021
9999    0.000062
Name: registration_year, Length: 95, dtype: float64


We can try to find out the number of cars registered between 1900-2016 and then divide that by the total number of cars, which would give us a percentage of cars registered outside of a reasonable timeframe.

In [20]:
#the ~ negates the argument. So it finds the number that is between 1900-2016, and then NOTs it
print((~autos["registration_year"].between(1900,2016)).sum() / autos.shape[0],'not in range')
print((autos["registration_year"].between(1900,2016)).sum() / autos.shape[0],'in range')

0.038793369710697 not in range
0.961206630289303 in range


So less than 4% of our data falls outside this reasonable time frame, so we can safely remove them.

In [21]:
#removes the violating rows using boolean masks via pandas methods
autos = autos[autos['registration_year'].between(1900,2016)] #'between()' is the boolean mask

#sorts by year, most recent first
print((autos['registration_year']
       .value_counts(normalize=True, dropna=False)
       .sort_index(ascending=False)
       .head(10)))

2016    0.026135
2015    0.008397
2014    0.014203
2013    0.017202
2012    0.028063
2011    0.034768
2010    0.034040
2009    0.044665
2008    0.047450
2007    0.048778
Name: registration_year, dtype: float64


In [22]:
#sorts by perfecntage, largest first
print(autos['registration_year']
     .value_counts(normalize=True, dropna=False)
     .sort_values(ascending=False)
     .head(10))

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
Name: registration_year, dtype: float64


As we can now see above, these are fairly uniform distributions for the different registration years. Most of the vehicles in our data set were registered in the past ~20 years.


## Exploring Price by Brand

Now we will explore the brand column a little bit.

In [23]:
autos["brand"].value_counts(normalize=True).head(10)

volkswagen       0.211264
bmw              0.110045
opel             0.107581
mercedes_benz    0.096463
audi             0.086566
ford             0.069900
renault          0.047150
peugeot          0.029841
fiat             0.025642
seat             0.018273
Name: brand, dtype: float64

German manufacturers represent four out of the top five brands, almost 50% of the overall listings. Volkswagen is by far the most popular brand, with approximately double the cars for sale of the next two brands combined.

There are lots of brands that don't have a significant percentage of listings, so we will limit our analysis to brands representing more than 5% of total listings.

In [24]:
brand_counts = autos["brand"].value_counts(normalize=True)
common_brands = brand_counts[brand_counts > .05].index
print(common_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [25]:
brand_mean_prices = {}
print('{:^}'.format('Most Common Brands Mean Price'))

for brand in common_brands:
    #goes through ALL the autos data and checks to see the brands match
    brand_only = autos[autos["brand"] == brand]
    mean_price = brand_only["price"].mean()
    brand_mean_prices[brand] = int(mean_price)
    print('{:>14}: {:<}'.format(brand,brand_mean_price[brand]))

Most Common Brands Mean Price


NameError: name 'brand_mean_price' is not defined

Of the top 5 brands, there is a fairly distinct price gap. Opel and Ford are the most affordable (i.e. cheapest) to buy, while BMW, Mercedes Benz, and Audi are the most expensive. The most commonly listed brand, Volkswagen, is right in the middle, which might have something to do with its popularity. It's the best of both worlds it seems.

In [ ]:
brand_mean_mileage = {}
print('{:^}'.format('Most Common Brands Mean Mileage'))

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_mileage = brand_only["odometer_km"].mean()
    brand_mean_mileage[brand] = int(mean_mileage)
    print('{:>14}: {:<}'.format(brand,brand_mean_mileage[brand]))


We can see that all six of the top brands have  very similar mean mileage counts, with the lowest being 124k, and the highest being 132k.

In [ ]:
bmp_series = pd.Series(brand_mean_prices)
bmm_series = pd.Series(brand_mean_mileage)
pd.DataFrame(bmp_series, columns=['mean_price'])

In [ ]:
#this takes our series and sorts them for improved readability
mean_mileage = pd.Series(bmm_series).sort_values(ascending=False)
mean_prices = pd.Series(bmp_series).sort_values(ascending=False)

brand_info = pd.DataFrame(mean_mileage,columns=['mean_mileage'])
brand_info

In [ ]:
brand_info['mean_price'] = mean_prices
brand_info

The range of car mileages does not vary as much as the prices do by brand, instead all falling within 10% for the top brands. There is a slight trend to the more expensive vehicles having higher mileage, with the less expensive vehicles having lower mileage.

## Extra / Miscellaneous Testing
Here I will try to further clean up the data to practice my programming skills. First, I will try to translate as many German words in our dataset as possible, starting with the gearbox column.
### Translating German words

In [ ]:
autos

In [ ]:
german_list = ['vehicle_type', 'gearbox', 'fuel_type', 'unrepaired_damage']
for p in german_list:
    print(str(p), '\n', pd.unique(autos[p]), '\n')

In [ ]:
autos['vehicle_type'] = (autos['vehicle_type']
                         .map({'kleinwagen' : 'compact',
                               'andere': 'other',
                               'kombi': 'station wagon',
                               'cabrio': 'convertible'},
                              na_action='ignore')
                       )
# replace NaN to string 'no information':
autos['vehicle_type'] = autos['vehicle_type'].fillna('not listed')

autos['gearbox'] = (autos['gearbox']
                    .map({'manuell': 'manual',
                          'automatik': 'automatic'},
                         na_action='ignore')
                   )
# replace NaN to string 'no information':
autos['gearbox'] = autos['gearbox'].fillna('not listed')

autos['fuel_type'] = (autos['fuel_type']
                      .map({'benzin': 'petrol',
                            'elektro' : 'electro',
                            'andere': 'other'},
                         na_action='ignore')
                   )
# replace NaN to string 'no information':
autos['fuel_type'] = autos['fuel_type'].fillna('not listed')

autos['unrepaired_damage'] = (autos['unrepaired_damage']
                              .map({'nein': 'no',
                                    'ja': 'yes'})
                           )
# replace NaN to string 'no information':
autos['unrepaired_damage'] = autos['unrepaired_damage'].fillna('not listed')

In [ ]:
autos

In [ ]:
#to verify the words have been translated, run this again
german_list = ['vehicle_type', 'gearbox', 'fuel_type', 'unrepaired_damage']
for p in german_list:
    print(str(p), '\n', pd.unique(autos[p]), '\n', '\n')

### Converting Dates to Numeric Values
Now I will try to clean up the date format a bit. I will convert it from a string "2016-03-12" to an int 20160312.

In [ ]:
autos[["date_crawled","ad_created","last_seen"]][0:5]

In [ ]:
#splites the date_crawled column into two strings: date and time
autos['date_crawled: time'] = autos['date_crawled'].str.split().str[1]
autos['date_crawled'] = (autos['date_crawled']
                         .str.replace('-', '')
                         .str.split().str[0].astype(int)
                        )

autos['ad_created: time'] = autos['ad_created'].str.split().str[1]
# because there is no data in " ad_created: time", we will drop this column.
autos.drop('ad_created: time', inplace=True, axis=1)

autos['ad_created'] = (autos['ad_created'].str.replace('-', '')
                       .str.split().str[0].astype(int)
                      )

autos['last_seen: time'] = autos['last_seen'].str.split().str[1]
autos['last_seen'] = (autos['last_seen']
                      .str.replace('-', '')
                      .str.split().str[0].astype(int)
                     )
autos[["date_crawled","ad_created","last_seen"]][0:5]

### Most common model per brand?

In [ ]:
autos['brand'].value_counts().head(10)

In [ ]:
common_model = {}

for c in autos['brand'].value_counts().index:
    brands = autos[autos['brand'] == c]
    desc = brands['name'].describe()
    common_model[c] = desc[2]
    
print('Most common model for each brand:', '\n')
for key in common_model:
    print('{:>14}: {:<}'.format(key,common_model[key]))

### Prices vs. Mileage - Is There a Connection?

In [ ]:
autos['odometer_km'].value_counts().sort_index(ascending=False)

Below, we will find the mean odometer value for each brand.

In [ ]:
all_brands = autos['brand'].value_counts().index
brand_mean_odo = {}
print('{:>14}: {:>6}'.format('Brand','Mean_Odo'))

for brand in all_brands:
    brand_only = autos[autos['brand'] == brand]
    mean_odo = brand_only['odometer_km'].mean()
    brand_mean_odo[brand] = int(mean_odo)
    print('{:>14}: {:>6d}'.format(brand,brand_mean_odo[brand]))

Now we will do the same, but for the price of each brand.

In [ ]:
brand_mean_price = {}
print('{:>14}: {:>6}'.format('Brand','Mean_Price'))

for brand in all_brands:
    brand_only = autos[autos['brand'] == brand]
    mean_price = brand_only['price'].mean()
    brand_mean_price[brand] = int(mean_price)
    print('{:>14}: {:>6d}'.format(brand,brand_mean_price[brand]))

Now we will create a new DataFrame for each brand_mean_odo and brand_mean_price, and combine the two in one. First we need to create a Series using the pd.Series constructor.

In [ ]:
odo_series = pd.Series(brand_mean_odo).sort_values(ascending=False)
price_series = pd.Series(brand_mean_price).sort_values(ascending=False)

Now we will take both series and combine them into a new DataFrame using the pd.DataFrame constructor.

In [ ]:
odo_vs_price = pd.DataFrame(odo_series, columns=['mean_odo']) #sorts by odometer_km highest->lowest
odo_vs_price['mean_price'] = price_series

price_vs_odo = pd.DataFrame(price_series, columns=['mean_price']) #sorts by price highest->lowest
price_vs_odo['mean_odo'] = odo_series

In [ ]:
# function that print first and last 4 rows:
def head_tail(df, x):
    return df.head(x).append(df.tail(x))

In [ ]:
print('First 8 and last 8, sorted by odometer (highest -> lowest)')
print(head_tail(odo_vs_price, 8))

In [ ]:
print('First 8 and last 8, sorted by price (highest -> lowest)')
print(head_tail(price_vs_odo, 8))

We can see that Trabant has the lowest mileage for the lowest price. However, Daihatsu, Rover, and Daewoo all have lower prices than Trabant, while having more than double the mileage. Porsche is (obviously) the most expensive in our data (after we removed the outlier prices of course). Of the cars that cost over 10,000, Jeep has the highest mileage. Lada is the next best in terms of low mileage/price.

### Damaged vs Undamaged - How is price affected?

In [ ]:
#mean price of all the cars in dataset
mean_price_all = autos['price'].mean()

#mean price of cars that have damage
damaged = autos[autos['unrepaired_damage'] == 'yes']
mean_price_damaged = damaged['price'].mean()

#mean price of cars without any damage
undamaged = autos[autos['unrepaired_damage'] == 'no']
mean_price_undamaged = undamaged['price'].mean()

print('{:^22}'.format('Mean Prices of Cars'))
print('{:>10}: {:>5}'.format('All',int(mean_price_all)))
print('{:>10}: {:>5}'.format('Damaged',int(mean_price_damaged)))
print('{:>10}: {:>5}'.format('Undamaged',int(mean_price_undamaged)))

As we can see above, the price of a damaged car is significantly lower than undamaged cars. On average an undamaged car is almost $5000 cheaper than a damaged one. That's quite significant I think. This would make sense because people don't typically want to buy damaged goods.

## Conclusion

We have utilized pandas and NumPy to quickly and easily clean, sort, and manipulate a large set of data. There were many things that required cleaning, as well as things that were technically fine, but needed to be changed for improved readability. We made heavy use of DataFrames in this project, as well as boolean index masking and aggregation. 

In terms of cars, it should be no surprise that there are usually three "tiers": cheap, average, and expensive. Some cheap cars have low mileage, for example Tabant and Lada. Some expensive cars such as Sonstige Autos and Mini also have rather decent mileage for their price tag. As mentioned earlier there is a very slight trend to the more expensive vehicles having higher mileage, with the less expensive vehicles having lower mileage.

There are other factors at play as well. Obviously really famous and popular brands such as Porsche will typically be more expensive and overpriced. Damaged cars will typically be cheaper as well.